File: AUN-QA.IdentityService/Services/SystemGroup/SystemGroupService.cs
using AUN_QA.IdentityService.DTOs.Base;
using AUN_QA.IdentityService.DTOs.CoreFeature.SystemGroup.Dtos;
using AUN_QA.IdentityService.DTOs.CoreFeature.SystemGroup.Requests;
using AUN_QA.IdentityService.Helpers;
using AUN_QA.IdentityService.Infrastructure.Data;
using AutoDependencyRegistration.Attributes;
using AutoMapper;
using Npgsql;

namespace AUN_QA.IdentityService.Services.SystemGroup
{
    [RegisterClassAsTransient]
    public class SystemGroupService : ISystemGroupService
    {
        private readonly IdentityContext _context;
        private readonly IMapper _mapper;
        private readonly IHttpContextAccessor _contextAccessor;

        public SystemGroupService(
            IdentityContext context,
            IMapper mapper,
            IHttpContextAccessor contextAccessor)
        {
            _context = context;
            _mapper = mapper;
            _contextAccessor = contextAccessor;
        }

        public ModelSystemGroup GetById(GetByIdRequest request)
        {
            var data = _context.SystemGroups.Find(request.Id);
            if (data == null)
            {
                throw new Exception("Not found");
            }

            return _mapper.Map<ModelSystemGroup>(data);
        }

        public ModelSystemGroup Insert(SystemGroupRequest request)
        {
            var data = _context.SystemGroups.Where(x =>
                x.Name == request.Name
                && !x.IsDeleted
            );

            if (data.Any())
            {
                throw new Exception("Tên nhóm đã tồn tại");
            }

            var add = _mapper.Map<Entities.SystemGroup>(request);
            add.Id = request.Id == Guid.Empty ? Guid.NewGuid() : request.Id;
            add.CreatedBy = _contextAccessor.HttpContext?.User?.Identity?.Name ?? "System";
            add.CreatedAt = DateTime.Now;

            _context.SystemGroups.Add(add);
            _context.SaveChanges();

            return _mapper.Map<ModelSystemGroup>(add);
        }

        public ModelSystemGroup Update(SystemGroupRequest request)
        {
            var data = _context.SystemGroups.Where(x =>
                x.Name == request.Name
                && !x.IsDeleted && x.Id != request.Id);

            if (data.Any())
            {
                throw new Exception("Tên nhóm đã tồn tại");
            }

            var update = _context.SystemGroups.Find(request.Id);
            if (update == null)
            {
                throw new Exception("Dữ liệu không tồn tại");
            }

            _mapper.Map(request, update);

            update.UpdatedBy = _contextAccessor.HttpContext?.User?.Identity?.Name ?? "System";
            update.UpdatedAt = DateTime.Now;
            _context.SystemGroups.Update(update);
            _context.SaveChanges();

            return _mapper.Map<ModelSystemGroup>(update);
        }

        public string DeleteList(DeleteListRequest request)
        {
            foreach (var id in request.Ids)
            {
                var delete = _context.SystemGroups.Find(id);
                if (delete == null)
                {
                    throw new Exception("Dữ liệu không tồn tại");
                }

                delete.IsDeleted = true;
                delete.UpdatedBy = _contextAccessor.HttpContext?.User?.Identity?.Name ?? "System";
                delete.UpdatedAt = DateTime.Now;

                _context.SystemGroups.Update(delete);
            }

            _context.SaveChanges();
            return String.Join(',', request.Ids);
        }

        public async Task<GetListPagingResponse<ModelSystemGroupGetListPaging>> GetList(GetListPagingRequest request)
        {
            var parameters = new[]
            {
                new NpgsqlParameter("i_textsearch", request.TextSearch),
                new NpgsqlParameter("i_pageindex", request.PageIndex - 1),
                new NpgsqlParameter("i_pagesize", request.PageSize),
            };

            var result = await _context.ExcuteFunction<GetListPagingResponse<ModelSystemGroupGetListPaging>>("fn_system_group_getlistpaging", parameters);
            return result;
        }

        public List<MODELCombobox> GetAllForCombobox()
        {
            var data = _context.SystemGroups.Where(x => !x.IsDeleted && x.IsActived).ToList();
            var result = data.Select(x => new MODELCombobox
            {
                Text = x.Name,
                Value = x.Id.ToString(),
                Parent = x.ParentId.HasValue ? data.FirstOrDefault(y => y.Id == x.ParentId)?.Name : ""
            }).OrderBy(x => x.Text).ToList();

            return result;
        }
    }
}

File: AUN-QA.Web/src/services/system/systemGroup.service.ts
import api, { type ApiResponse } from "@/lib/api";
import type { SystemGroup, SystemGroupGetListPaging } from "@/types/system/systemGroup.types";
import type { GetListPagingRequest, ModelCombobox } from "@/types/base/base.types";
import { API_ENDPOINTS } from "@/config/constants";
import type { GetListPagingResponse } from "@/types/base/base.types";

export const systemGroupService = {
  getList: async (request: GetListPagingRequest): Promise<ApiResponse<GetListPagingResponse<SystemGroupGetListPaging>>> => {
    return api.post<GetListPagingResponse<SystemGroupGetListPaging>>(API_ENDPOINTS.System.SystemGroup.GET_LIST, request);
  },

  getById: async (id: string): Promise<ApiResponse<SystemGroup>> => {
    return api.get<SystemGroup>(API_ENDPOINTS.System.SystemGroup.GET_BY_ID, { params: { id } });
  },

  insert: async (data: SystemGroup): Promise<ApiResponse<SystemGroup>> => {
    return api.post<SystemGroup>(API_ENDPOINTS.System.SystemGroup.INSERT, data);
  },

  update: async (data: SystemGroup): Promise<ApiResponse<SystemGroup>> => {
    return api.put<SystemGroup>(API_ENDPOINTS.System.SystemGroup.UPDATE, data);
  },

  deleteList: async (ids: string[]): Promise<ApiResponse<SystemGroup[]>> => {
    return api.delete<SystemGroup[]>(API_ENDPOINTS.System.SystemGroup.DELETE_LIST, { data: { ids } });
  },

  getAllCombobox: async (): Promise<ApiResponse<ModelCombobox[]>> => {
    return api.get<ModelCombobox[]>(API_ENDPOINTS.System.SystemGroup.GET_ALL_COMBOBOX);
  },
};
